### PR TITLE
Performance Optimization 

### PR DESCRIPTION
**Pull request by RJBornToWin**
# Description
Code changes to improve the front end application performance.
Made analysis on the code flow of Timelog, Summary Bar and Leaderboard component to identify candidates causing delayed loading of Dashboard/ Timelog pages.

## Mainly changes explained:
1. Remove console.log statements wherever possible while working on Dashboard and Timelog performance.
2. Import only required functions from "lodash" instead of entire library
3. Add overflow scroll to badges container as it overflows outside of the container.
4. Remove unused imports and variables
5. Lazy loaded badge images using HTML5 feature.
6. SummaryBar: useEffect has been updated to FETCH data from backend only if the requested page's "userId" is different from the authenticated user profile which is stored in redux global state. Else the already stored data is displayed.
7. SummaryBar: Moved statements from "return" to "useEffect()" wherever possible

## How to test:
check into current branch
do `npm install` and `...` to run this PR locally
Log in with differet role users
Check for general functioning of Dashboard, Timelog and Leaderboard.

## Note:
All the changes are a step closer to cleaner code and best practices. With multiple tests no considerable amount of performance improvement has been achieved. During this process, I have 3 more open points which when addressed will contribute for considerable performance. Will document the open points in the required docs.

### CODE REVIEW

**Comment by Lunaliar** [comment]
Hey @RJBornToWin, to me this change is night and day! I don't have reliable internet sometimes when I'm at my other job so at times I would be waiting up to a minute for different pages to load which really cut into the amount of time I was able to work on the app. I read through your code to double-check the changes made and I didn't see anything that raised any red flags, amazing attention to detail. I also tried to find any leftover instances of a full lodash import and was unable to come across any. This is a game-changer great job!

**Review by Lunaliar — DISMISSED** [review]
Great job!

**Review by Lucas-E — COMMENTED** [review]
Hey RJ. I've tested your PR and it's awesome. It's a good step towards clean code. I've seen that the badge images inside the profile are not set to lazy loading, the component can be found in UserProfile/BadgeImage.jsx. Again, this work is super awesome!
Captura de tela de 2023-02-28 14-29-12
Captura de tela de 2023-02-28 14-30-20

**Comment by mpavithr** [comment]
Hi, I tested this by logging in as an admin account and volunteer account and I didn't see any issues with the dashboard.
When it comes to removing console.log statements, I see there are some commented-out console.log statements in the code remaining such as in some files under HighestGoodNetworkApp/src/actions and other such files. I am not sure if you wanted to remove all of these, but just wanted you to know!

**Comment by crystalsong2021** [comment]
Thank you for refactoring the code towards cleaner and make use of the lazy loading function. I am still new and getting use to the code. When I test out the application, I noticed the application frozen after I clicked the **project** from the main menu.
https://user-images.githubusercontent.com/18197013/222006817-2edf2ee2-a2af-444a-a6c3-9669e795f576.mov

File: src/actions/team.js [hunk 1]
```diff
@@ -50,7 +50,7 @@ export const fetchAllManagingTeams = (userId, managingTeams) => {
       allMembers = allMembers.concat(data[i].data);
     }
     console.log('allManagingTeams:', allManagingTeams);
```
Comment (crystalsong2021) [inline src/actions/team.js:52]: console.log removal in this line

File: src/utils/effortColors.js [hunk 2]
```diff
@@ -1,64 +1,64 @@
...
 
 //For progress bar that shows the percentage of the completion
 //Set 'invert' to true when used for the bars of TASK
-export const getProgressColor = (effort,commit,invert=false) => {
+export const getProgressColor = (effort, commit, invert = false) => {
   let color = 'white';
   let percentage = 0;
-  if (commit>0){
-    percentage = Math.round(effort*100/commit);
+  if (commit > 0) {
```
Comment (crystalsong2021) [inline src/utils/effortColors.js:8]: Nice job on fixing the space. Thank you for your keen eye.

**Comment by anant2047** [comment]
This is really a great job! I went through the code thoroughly, didn't find any issues, and tested a lot of scenarios. Can see the performance going up. Bravo!

I do see there is no more console logging and sharing the screen recording for the same.

In development branch there is a lot of logging - 

https://user-images.githubusercontent.com/97596031/222048689-ae9deb96-684c-4efd-818c-4756dff816df.mov

All of that is removed in your changes and will help in the performance substantially -

https://user-images.githubusercontent.com/97596031/222048751-4e180d9b-d854-46d4-94a1-9d6573bafa1f.mov

**Review by anant2047 — APPROVED** [review]

File: src/utils/effortColors.js (hunk 2 above)
Comment (RJBornToWin) [reply src/utils/effortColors.js:8]: Thanks for reviewing. FYI, if we have prettier configured, it auto corrects code spacing on file save.

File: src/actions/team.js (hunk 1 above)
Comment (RJBornToWin) [reply src/actions/team.js:52]: Thanks for reviewing. I have not removed all console.log instances. This is a big task and has to be achieved over time. But only those that I found while working on dashboard and timelog page performance.

**Review by NyahJ — DISMISSED** [review]
Hi @RJBornToWin,

I have followed your suggestion for testing, using different accounts to test all the functional features, they all worked as intended.
As for the codes part, they are clear and clean. Thanks!

截屏2023-03-01 下午12 35 29
截屏2023-03-01 下午12 36 46

**Review by vitorsaa2k — DISMISSED** [review]
The perfomace really increased, impressive! Great job, what a good PR.

I've found a bug when, logging out of a volunteer account and logging into an admin account this error appear:

image

It also occour on the development branch so, it isn't from this PR. 

Video triggering the error:

https://user-images.githubusercontent.com/17365161/222242559-97b94d98-2a8e-4845-b8cd-4b26bc7c4302.mp4

I could solve the bug by adding those lines in the SummaryBar.jsx file (after line 87):

image

After solving the bug:

https://user-images.githubusercontent.com/17365161/222258209-4eb9650d-537d-4be5-85c3-5a53fb86951d.mp4

despite this bug everything else is working fine, so I'll approve this PR. But fix the bug if possible please and let me know when you do it.

**Review by crystalsong2021 — DISMISSED** [review]

**Comment by angelinatr15** [comment]
Hi @RJBornToWin, 

Everything is working as expected. Thank you for removing the unneeded console.log(), imports, and variables. It makes the code look much cleaner and minimizes unnecessary console.log() time. I like that you only import needed functions from lodash which saves a lot of memory. The change to the useEffect() is great because this way the client would not have to fetch data from the server if they are already logged into that userId. Excellent job!

**Review by angelinatr15 — DISMISSED** [review]

**Comment by RJBornToWin** [comment]
Thanks to all the reviewers! 
@vitorsaa2k  I have made the suggested change.
@Lucas-E I am not doing the lazy loading change in user profile because: 1. The image shows up on the top of the page. i.e needs to be visible immediately on page load. But Lazy loading it will cause it to load at the last and might add to the glitchy rendering. Whereas the badges section is hidden from user view on page load in the Dashboard/Timelog pages 2. There are only 5 images fetched at max. This page is comparatively lighter than Dashboard/Timelog pages.

**Review by kamtoeddy — COMMENTED** [review]
Hi @RJBornToWin,

Great job improving the codebase

I did see some errors in the console when testing with both Admin & Volunteer accounts.
Here are the errors with my proposed solutions

1.
Screenshot (547)
Screenshot (549)

2. 
Screenshot (550)
Screenshot (551)

3.
Screenshot (553)
Screenshot (552)

Please if these are not related to your PR, let me know so I can approve

**Review by vitorsaa2k — DISMISSED** [review]
Tested and no more bugs now. PR approved

https://user-images.githubusercontent.com/17365161/222493867-15e47256-743b-4026-a4a0-d47f7861bda1.mp4

**Review by kangjay88 — APPROVED** [review]
Hi @RJBornToWin,

Reviewed the PR changes and codebase looks cleaner, tested dashboard/time functionality with volunteer and admin roles. No issues, less console logs, and increased performance! Thank you

**Comment by agujjar2206** [comment]
Hello @RJBornToWin, thank you for submitting this PR. I tested it out and everything seems to be working well. The changes made to the Timelog, Summary Bar, and Leaderboard components have improved the front-end application's performance. I appreciate the effort put into removing console.log statements, importing only required functions, adding overflow scroll, removing unused imports and variables, and lazy loading badge images. Additionally, the updated useEffect in SummaryBar to fetch data from the backend only when necessary is a great optimization. Overall, this is a step towards cleaner code and best practices. Looking forward to seeing the open points addressed as well!

Thank you.

**Review by agujjar2206 — DISMISSED** [review]

File: src/components/Timelog/Timelog.jsx [hunk 3]
```diff
@@ -100,24 +100,28 @@ class Timelog extends Component {
   state = this.initialState;
 
   async componentDidMount() {
-    const userId = this.props.auth.user.userid; //It was refering to a undefined userid, so i got the id from the store to the state of the component.
-    await this.props.getUserProfile(userId);
+    const userId =
+      this.props?.match?.params?.userId || this.props.asUser || this.props.auth.user.userid; //Including fix for "undefined"
```
Comment (RJBornToWin) [inline src/components/Timelog/Timelog.jsx:104]: Including this from development that fixed undefined user id for timelog page.

File: src/components/SummaryBar/SummaryBar.jsx [hunk 4]
```diff
@@ -24,31 +24,26 @@ import badges_icon from './badges_icon.png';
...
   const [userProfile, setUserProfile] = useState(undefined);
+  const [infringements, setInfringements] = useState(0);
+  const [badges, setBadges] = useState(0);
+  const [totalEffort, setTotalEffort] = useState(0);
+  const [weeklySummary, setWeeklySummary] = useState([]);
+
   const [tasks, setTasks] = useState(undefined);
   const authenticateUser = useSelector(state => state.auth.user);
+  const gsUserprofile = useSelector(state => state.userProfile);
+  const gsUserTasks = useSelector(state => state.userTask);
```
Comment (RafaelCastilhoCastro) [inline src/components/SummaryBar/SummaryBar.jsx:43]: @RJBornToWin You think is possible to rename these two "gs" variables to improve readability?

File: src/components/SummaryBar/SummaryBar.jsx [hunk 5]
```diff
@@ -178,16 +171,12 @@ const SummaryBar = props => {
...
   if (userProfile !== undefined && leaderData !== undefined) {
-    const infringements = getInfringements(userProfile);
-    const badges = getBadges(userProfile);
-    const { firstName, lastName, email, _id } = userProfile;
-    let totalEffort = parseFloat(leaderData.find(x => x.personId === asUser).tangibletime);
-    const weeklyCommittedHours = userProfile.weeklyCommittedHours
-      ? userProfile.weeklyCommittedHours
+    const weeklyCommittedHours = userProfile.weeklycommittedHours
+      ? userProfile.weeklycommittedHours
       : userProfile.weeklyComittedHours;
```
Comment (RafaelCastilhoCastro) [inline src/components/SummaryBar/SummaryBar.jsx:176]: @RJBornToWin This block should stay like it is? There's a correct "weeklycommittedHours" and a misspelled "weeklyComittedHours" prop being used together. This was already on the code before your refactoring process btw.

File: src/components/SummaryBar/SummaryBar.jsx [hunk 6]
```diff
@@ -178,16 +171,12 @@ const SummaryBar = props => {
   };
 
   if (userProfile !== undefined && leaderData !== undefined) {
-    const infringements = getInfringements(userProfile);
-    const badges = getBadges(userProfile);
-    const { firstName, lastName, email, _id } = userProfile;
-    let totalEffort = parseFloat(leaderData.find(x => x.personId === asUser).tangibletime);
-    const weeklyCommittedHours = userProfile.weeklyCommittedHours
-      ? userProfile.weeklyCommittedHours
+    const weeklyCommittedHours = userProfile.weeklycommittedHours
```
Comment (RafaelCastilhoCastro) [inline src/components/SummaryBar/SummaryBar.jsx:174]: @RJBornToWin This block should stay like it is? There's a correct "weeklycommittedHours" and a misspelled "weeklyComittedHours" prop being used together. This was already on the code before your refactoring process btw.

**Review by RafaelCastilhoCastro — APPROVED** [review]
Hi @RJBornToWin,
I tested your PR and the performance did improve. All highlighted features worked as intended. Excellent job.
I'm approving the PR. When possible please address my comments so we can ask Jae to merge it. Thanks.

**Pull request merged**